### PR TITLE
Upgrades urbanopt

### PR DESCRIPTION
**Pull request by ann-sherin**
Addresses issues #113 , #114 , #115 , #120,  #122, #126 , #127 , #137

### CODE REVIEW

File: disco/models/upgrade_cost_analysis_generic_input_model.py [hunk 1]
```diff
@@ -0,0 +1,1166 @@
...
+        title="r0",
+        description="r0",
+        upgrade_option_property=True,
+    )
+    x0: Any = Field(
+        title="x0",
+        description="x0",
+        upgrade_option_property=True,
+    )
+    C1: Any = Field(
```
Comment (daniel-thom) [inline disco/models/upgrade_cost_analysis_generic_input_model.py:641]: It's more predictable and straightforward if you always use lower-case, snake_case. There is quite a bit of inconsistency in this file.

File: disco/models/upgrade_cost_analysis_generic_output_model.py [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+    )
+    
+
+class ExtendedEnum(enum.Enum):
+    @classmethod
+    def list(cls):
+        return list(map(lambda c: c.value, cls))
+    
+
+class capacitor_controller_output_type(ExtendedEnum):
```
Comment (daniel-thom) [inline disco/models/upgrade_cost_analysis_generic_output_model.py:499]: Use PascalCase for class names.

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 3]
```diff
@@ -362,12 +388,189 @@ def get_circuit_info():
     if len(all_xfmr_df.loc[all_xfmr_df["substation_xfmr_flag"] == True]) > 0:
         data_dict["substation_xfmr"] = all_xfmr_df.loc[all_xfmr_df["substation_xfmr_flag"] ==
                                                        True].to_dict(orient='records')[0]
-        data_dict["substation_xfmr"]["kVs"] = ast.literal_eval(data_dict["substation_xfmr"]["kVs"])
         # this checks if the voltage kVs are the same for the substation transformer
         data_dict["substation_xfmr"]["is_autotransformer_flag"] = len(set(data_dict["substation_xfmr"]["kVs"])) <= 1
     return data_dict
 
 
+def _create_thermal_output_summary_(all_original_equipment, all_latest_equipment, thermal_equipment_type_list,
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:396]: Fixed

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 4]
```diff
@@ -1347,21 +1351,20 @@ def determine_available_line_upgrades(line_loading_df):
     """This function creates a dataframe of available line upgrades by dropping duplicates from line dataframe passed.
     """
     property_list =  _extract_specific_model_properties_(model_name=LineCatalogModel, field_type_key="determine_upgrade_option", field_type_value=True)
+    line_loading_df["kV"] = line_loading_df["kV"].round(5)
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/common_functions.py:1375]: This is to make sure the kV precision matches from external catalog / internal catalog match. It was there before as well, I just moved it higher in the function

File: disco/extensions/upgrade_simulation/upgrades/automated_thermal_upgrades.py [hunk 5]
```diff
@@ -248,35 +261,37 @@ def determine_thermal_upgrades(
...
     else:
         feeder_stats = {}
-    feeder_stats["stage_results"].append( get_upgrade_stage_stats(dss, upgrade_stage="Final", upgrade_type="thermal", xfmr_loading_df=xfmr_loading_df, line_loading_df=line_loading_df, 
-                                        bus_voltages_df=bus_voltages_df) )
-    dump_data(feeder_stats, feeder_stats_json_file, indent=4)
+    regcontrols_df = get_regcontrol_info(correct_PT_ratio=False)
+    capacitors_df = get_capacitor_info(correct_PT_ratio=False)
+    feeder_stats["stage_results"].append( get_upgrade_stage_stats(dss, upgrade_stage="final", upgrade_type="thermal", xfmr_loading_df=xfmr_loading_df, line_loading_df=line_loading_df, 
+                                        bus_voltages_df=bus_voltages_df, capacitors_df=capacitors_df, regcontrols_df=regcontrols_df) )
+    dump_data(feeder_stats, feeder_stats_json_file, indent=2)
```
[the remark this answers is not in the document]
Comment (ann-sherin) [reply disco/extensions/upgrade_simulation/upgrades/automated_thermal_upgrades.py:268]: Leaving it as is, since consistent with other models.

**Review by daniel-thom — APPROVED** [review]

**Pull request merged**